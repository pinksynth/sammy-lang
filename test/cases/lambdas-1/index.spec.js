const {
  performSnapshotAssertions,
  expectCompiledEval,
} = require("../../helpers")
performSnapshotAssertions("Lambda definitions", __dirname)

test("Lambda implementation", () => {
  expectCompiledEval(`
		multiply = @ a b { a * b }
		divide = @{ $1 / $2 }

		[
			multiply(2 3)
			divide(11 2)
		]
	`).toEqual([6, 5.5])
})
