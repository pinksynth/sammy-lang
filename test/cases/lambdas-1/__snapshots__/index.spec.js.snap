// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ“¸ Lambda definitions 1`] = `
[
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 1,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "multiply",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 10,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "ASSIGNMENT",
    "value": "=",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 13,
    "columnNumberStart": 12,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "LAMBDA_OPEN",
    "value": "@",
  },
  {
    "columnNumberEnd": 14,
    "columnNumberStart": 13,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 15,
    "columnNumberStart": 14,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 16,
    "columnNumberStart": 15,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 17,
    "columnNumberStart": 16,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "b",
  },
  {
    "columnNumberEnd": 18,
    "columnNumberStart": 17,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 19,
    "columnNumberStart": 18,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "CURLY_OPEN",
    "value": "{",
  },
  {
    "columnNumberEnd": 20,
    "columnNumberStart": 19,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 21,
    "columnNumberStart": 20,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 22,
    "columnNumberStart": 21,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 23,
    "columnNumberStart": 22,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "OPERATOR_INFIX",
    "value": "*",
  },
  {
    "columnNumberEnd": 24,
    "columnNumberStart": 23,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 25,
    "columnNumberStart": 24,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "b",
  },
  {
    "columnNumberEnd": 26,
    "columnNumberStart": 25,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 27,
    "columnNumberStart": 26,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "CURLY_CLOSE",
    "value": "}",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 27,
    "lineNumberEnd": 2,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 7,
    "columnNumberStart": 1,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "VAR",
    "value": "divide",
  },
  {
    "columnNumberEnd": 8,
    "columnNumberStart": 7,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 8,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "ASSIGNMENT",
    "value": "=",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 10,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "LAMBDA_OPEN",
    "value": "@",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "CURLY_OPEN",
    "value": "{",
  },
  {
    "columnNumberEnd": 13,
    "columnNumberStart": 12,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 15,
    "columnNumberStart": 13,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "CONCISE_LAMBDA_ARGUMENT",
    "value": "$1",
  },
  {
    "columnNumberEnd": 16,
    "columnNumberStart": 15,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 17,
    "columnNumberStart": 16,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "OPERATOR_INFIX",
    "value": "/",
  },
  {
    "columnNumberEnd": 18,
    "columnNumberStart": 17,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 20,
    "columnNumberStart": 18,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "CONCISE_LAMBDA_ARGUMENT",
    "value": "$2",
  },
  {
    "columnNumberEnd": 21,
    "columnNumberStart": 20,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 22,
    "columnNumberStart": 21,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "CURLY_CLOSE",
    "value": "}",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 22,
    "lineNumberEnd": 3,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 11,
    "columnNumberStart": 1,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "VAR",
    "value": "double_all",
  },
  {
    "columnNumberEnd": 12,
    "columnNumberStart": 11,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 13,
    "columnNumberStart": 12,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "ASSIGNMENT",
    "value": "=",
  },
  {
    "columnNumberEnd": 14,
    "columnNumberStart": 13,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 15,
    "columnNumberStart": 14,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "LAMBDA_OPEN",
    "value": "@",
  },
  {
    "columnNumberEnd": 16,
    "columnNumberStart": 15,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 18,
    "columnNumberStart": 16,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "CONCISE_LAMBDA_ARGUMENT",
    "value": "$1",
  },
  {
    "columnNumberEnd": 19,
    "columnNumberStart": 18,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 20,
    "columnNumberStart": 19,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "CURLY_OPEN",
    "value": "{",
  },
  {
    "columnNumberEnd": 21,
    "columnNumberStart": 20,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 23,
    "columnNumberStart": 21,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "CONCISE_LAMBDA_ARGUMENT",
    "value": "$1",
  },
  {
    "columnNumberEnd": 24,
    "columnNumberStart": 23,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "DOT",
    "value": ".",
  },
  {
    "columnNumberEnd": 27,
    "columnNumberStart": 24,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "VAR",
    "value": "map",
  },
  {
    "columnNumberEnd": 28,
    "columnNumberStart": 27,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "PAREN_OPEN",
    "value": "(",
  },
  {
    "columnNumberEnd": 29,
    "columnNumberStart": 28,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "LAMBDA_OPEN",
    "value": "@",
  },
  {
    "columnNumberEnd": 30,
    "columnNumberStart": 29,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "CURLY_OPEN",
    "value": "{",
  },
  {
    "columnNumberEnd": 31,
    "columnNumberStart": 30,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 33,
    "columnNumberStart": 31,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "CONCISE_LAMBDA_ARGUMENT",
    "value": "$1",
  },
  {
    "columnNumberEnd": 34,
    "columnNumberStart": 33,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 35,
    "columnNumberStart": 34,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "OPERATOR_INFIX",
    "value": "*",
  },
  {
    "columnNumberEnd": 36,
    "columnNumberStart": 35,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 37,
    "columnNumberStart": 36,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "NUMBER",
    "value": "2",
  },
  {
    "columnNumberEnd": 38,
    "columnNumberStart": 37,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 39,
    "columnNumberStart": 38,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "CURLY_CLOSE",
    "value": "}",
  },
  {
    "columnNumberEnd": 40,
    "columnNumberStart": 39,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "PAREN_CLOSE",
    "value": ")",
  },
  {
    "columnNumberEnd": 41,
    "columnNumberStart": 40,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 42,
    "columnNumberStart": 41,
    "lineNumberEnd": 3,
    "lineNumberStart": 3,
    "tokenType": "CURLY_CLOSE",
    "value": "}",
  },
]
`;

exports[`ðŸ“¸ Lambda definitions 2`] = `
{
  "children": [
    {
      "children": [
        {
          "args": [
            {
              "columnNumberStart": 14,
              "lineNumberStart": 1,
              "type": "IDENTIFIER",
              "value": "a",
            },
            {
              "columnNumberStart": 16,
              "lineNumberStart": 1,
              "type": "IDENTIFIER",
              "value": "b",
            },
          ],
          "children": [
            {
              "columnNumberStart": 22,
              "left": {
                "columnNumberStart": 20,
                "lineNumberStart": 1,
                "type": "IDENTIFIER",
                "value": "a",
              },
              "lineNumberStart": 1,
              "operator": "*",
              "right": {
                "columnNumberStart": 24,
                "lineNumberStart": 1,
                "type": "IDENTIFIER",
                "value": "b",
              },
              "type": "BINARY_EXPR",
            },
          ],
          "columnNumberStart": 12,
          "lineNumberStart": 1,
          "type": "LAMBDA",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 1,
      "type": "ASSIGNMENT",
      "variable": "multiply",
      "weak": false,
    },
    {
      "children": [
        {
          "args": [],
          "children": [
            {
              "columnNumberStart": 16,
              "left": {
                "columnNumberStart": 13,
                "lineNumberStart": 2,
                "type": "CONCISE_LAMBDA_ARGUMENT",
                "value": "$1",
              },
              "lineNumberStart": 2,
              "operator": "/",
              "right": {
                "columnNumberStart": 18,
                "lineNumberStart": 2,
                "type": "CONCISE_LAMBDA_ARGUMENT",
                "value": "$2",
              },
              "type": "BINARY_EXPR",
            },
          ],
          "columnNumberStart": 10,
          "lineNumberStart": 2,
          "type": "LAMBDA",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 2,
      "type": "ASSIGNMENT",
      "variable": "divide",
      "weak": false,
    },
    {
      "children": [
        {
          "args": [
            {
              "columnNumberStart": 16,
              "lineNumberStart": 3,
              "type": "CONCISE_LAMBDA_ARGUMENT",
              "value": "$1",
            },
          ],
          "children": [
            {
              "children": [
                {
                  "args": [],
                  "children": [
                    {
                      "columnNumberStart": 34,
                      "left": {
                        "columnNumberStart": 31,
                        "lineNumberStart": 3,
                        "type": "CONCISE_LAMBDA_ARGUMENT",
                        "value": "$1",
                      },
                      "lineNumberStart": 3,
                      "operator": "*",
                      "right": {
                        "columnNumberStart": 36,
                        "lineNumberStart": 3,
                        "type": "LITERAL_NUMBER",
                        "value": "2",
                      },
                      "type": "BINARY_EXPR",
                    },
                  ],
                  "columnNumberStart": 28,
                  "lineNumberStart": 3,
                  "type": "LAMBDA",
                },
              ],
              "columnNumberStart": 23,
              "function": {
                "columnNumberStart": 23,
                "left": {
                  "columnNumberStart": 21,
                  "lineNumberStart": 3,
                  "type": "CONCISE_LAMBDA_ARGUMENT",
                  "value": "$1",
                },
                "lineNumberStart": 3,
                "operator": ".",
                "right": {
                  "columnNumberStart": 24,
                  "lineNumberStart": 3,
                  "type": "IDENTIFIER",
                  "value": "map",
                },
                "type": "BINARY_EXPR",
              },
              "lineNumberStart": 3,
              "type": "FUNCTION_CALL",
            },
          ],
          "columnNumberStart": 14,
          "lineNumberStart": 3,
          "type": "LAMBDA",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 3,
      "type": "ASSIGNMENT",
      "variable": "double_all",
      "weak": false,
    },
  ],
  "type": "ROOT",
}
`;

exports[`ðŸ“¸ Lambda definitions 3`] = `"const multiply=(a,b)=>{return a*b};const divide=($1,$2)=>{return $1/$2};const double_all=($1)=>{return $1.map(($1)=>{return $1*2})}"`;
